### PR TITLE
feat(steam): add STEAM offline advantage modeling pipeline

### PR DESCRIPTION
**Pull request by LiuZhihao2022**
### Description

This PR introduces **STEAM**, a pair-classification value model for embodied
VLA RL, together with the offline pipeline that turns a trained STEAM critic
into per-frame advantages for CFG-RL training.

Instead of regressing a scalar return, STEAM scores an ordered frame pair
`(frame_t, frame_{t+k})` as **progress vs. regress**. Negative examples are
simply positive pairs fed in reverse order,
so a value signal can be learned from success demos plus rollouts without any
hand-labelled rewards.

**What's added**

- **STEAM model** (`rlinf/models/embodiment/steam/`): a SigLIP vision encoder +
  Gemma3 language backbone + MLP classification head. Per-frame image tokens are
  concatenated *in order* (not mean-pooled) so the model can tell `(t, t+k)`
  from `(t+k, t)`. Supports binary and multi-bin **signed-stride** classification,
  and a **deep ensemble** that aggregates members with the conservative
  **worst-of-N** rule (`A_STEAM = min_m A_m`) to suppress the overestimated
  advantages a single predictor assigns to out-of-distribution rollouts. The
  single-model and ensemble paths are cleanly separated (`modeling_steam.py`,
  `modeling_critic.py`, `ensemble_modeling_critic.py`).
- **Pair dataset** (`rlinf/data/datasets/steam/`): forward and regress pair dataset + collator
  that feeds a `[B, num_cameras, num_frames, 3, H, W]` tensor to the backbone,
  signed-stride binning with optional length-scaled stride
  (`binning.py`).
- **FSDP SFT worker** (`rlinf/workers/sft/fsdp_steam_sft_worker.py`): trains the
  STEAM critic (single or ensemble) on the RLinf FSDP SFT runner.
- **Offline advantage pipeline** (`examples/steam/process/`): computes per-frame
  `advantage_continuous` (signed score in `[-1, 1]`) from a trained ensemble
  STEAM critic, derives the boolean `advantage` label under either a `threshold`
  or per-pool `quantile` mode (independent `rollout_quantile` /
  `expert_quantile`), plus tools to relabel parquet under a new threshold,
  merge ensemble checkpoints, and visualize episodes/advantages.
- **Training entry + configs** (`examples/steam/value/`): `train_steam.py`,
  `run_steam_sft.sh`, and Hydra configs that resolve the shared FSDP
  `training_backend` from `examples/sft/config/` via `hydra.searchpath`.
- **Registration & validation** (`rlinf/config.py`, `rlinf/models/__init__.py`):
  register `steam` in `SupportedModel` / `EMBODIED_MODEL` / the model builder
  registry, with ensemble-settings validation in `validate_sft_cfg`.

**Supporting refactors (no behaviour change to existing models)**

- Extract a shared image+text processing base in
  `rlinf/models/embodiment/value_common/` (`BaseMultiViewImageProcessor` +
  `BaseValueTextProcessor`); `value_model` and `steam` become thin subclasses
  that only set their differing defaults. Public class names and
  `from_pretrained` paths are preserved, so existing checkpoints keep loading.
- Move the dependency-light, torchrun-based offline-annotation helpers to
  `rlinf/data/process/distributed.py` so the `compute_advantages*` scripts can
  run inside annotation venvs (where lerobot/megatron are unavailable), and
  dedup them out of `examples/recap/process/compute_advantages.py`.
- `fsdp_model_manager.py`: use fused AdamW when CUDA is available (graceful
  fallback) to cut optimizer warmup memory for ensemble models.
- Extend `rlinf/utils/logging.py` with helpers used by the STEAM worker/pipeline.

### Motivation and Context

Embodied RL with VLAs needs a dense, learnable value/advantage signal, but
collecting scalar reward labels is expensive. STEAM reframes value learning as
**pair classification over temporal order**: progress demos read forward,
regress demos read backward. This yields a self-supervised signal from success
demos and rollouts alone, and an ensemble gives a calibrated, uncertainty-aware
score that the offline pipeline turns into the `advantage` labels consumed by
the existing CFG-RL training path. The result is a drop-in value model for
embodied RL that does not depend on environment rewards.

### How has this been tested?

- **SFT training:** trained the STEAM critic end-to-end via
  `bash examples/steam/value/run_steam_sft.sh steam_value_model` on the FSDP
  runner in the ensemble configuration; verified the training loss and binary
  classification accuracy.
- **Offline advantage pipeline:** ran
  `examples/steam/process/run_compute_advantages_ensemble.sh` against a trained
  ensemble checkpoint to produce the `advantage_continuous` / `advantage`
  columns under the `quantile` label mode, and sanity-checked the outputs with
  `visualize_advantage.py`.
- **Downstream:** confirmed the produced advantages are consumed by the existing
  CFG-RL training path.

### Additional information (optional, e.g., figures and logs):

Training loss and accuracy
截屏2026-06-18 02 34 40

Prediction performance
episode_0000_summary

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:
- [x] My code follows the code style of this project. (ruff check/format applied)
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (not included in this PR)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/value_model/value_common/image_text_processing.py [hunk 1]
```diff
@@ -0,0 +1,691 @@
...
+    if images.dtype == torch.uint8:
+        resized_images = torch.round(resized_images).clamp(0, 255).to(torch.uint8)
+        constant_value: float = 0
+    elif images.dtype == torch.float32:
+        resized_images = resized_images.clamp(
+            float_clamp_range[0], float_clamp_range[1]
+        )
+        constant_value = float_pad_value
+    else:
+        raise ValueError(f"Unsupported image dtype: {images.dtype}")
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/models/embodiment/value_model/value_common/image_text_processing.py:120]: Resolved — uint8 is cast to float before `F.interpolate`, then rounded/clamped back to uint8.

File: rlinf/models/embodiment/value_model/steam/modeling_steam.py [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+        subtract ``image_mean`` and divide by ``image_std``. A safety
+        rescale/resize still handles degenerate inputs (e.g. raw uint8 or
+        out-of-range float tensors).
+        """
+        if flat_images.dtype == torch.uint8:
+            flat_images = flat_images.to(dtype=torch.float32) / 255.0
+        else:
+            flat_images = flat_images.to(dtype=torch.float32)
+            if bool(flat_images.max() > 1.0) or bool(flat_images.min() < 0.0):
+                flat_images = (flat_images / 255.0).clamp(0.0, 1.0)
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/models/embodiment/value_model/steam/modeling_steam.py:349]: Resolved — `[-1, 1]` inputs are now mapped via `(x+1)/2`; only `[0, 255]`-scaled data is divided by 255.

File: examples/steam/process/compute_advantages_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,982 @@
...
+    loader = DataLoader(
+        shard,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        prefetch_factor=prefetch_factor if num_workers > 0 else None,
+        persistent_workers=num_workers > 0,
+        pin_memory=True,
+        shuffle=False,
+        collate_fn=collate_fn,
+    )
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply examples/steam/process/compute_advantages_ensemble.py]: Resolved — `prefetch_factor`/`persistent_workers` are only passed when `num_workers > 0`.

File: rlinf/data/process/distributed.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        backend = dist_cfg.get("backend", "nccl")
+        timeout_seconds = dist_cfg.get("timeout", 1800)
+
+        if not dist.is_initialized():
+            dist.init_process_group(
+                backend=backend,
+                timeout=timedelta(seconds=timeout_seconds),
+            )
+
+        torch.cuda.set_device(local_rank)
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/data/process/distributed.py]: Resolved — `torch.cuda.set_device` is guarded by `torch.cuda.is_available()` and falls back to a cpu device.

File: examples/offline_rl/advantage_labeling/steam/run_steam_sft.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+mkdir -p "${LOG_DIR}"
+# Resolve shared config groups (training_backend/...) from examples/sft/config
+# regardless of what searchpath an individual config file declares.
+HYDRA_ARGS=(
+    "runner.logger.log_path=${LOG_DIR}"
+    "hydra.searchpath=[file://${SCRIPT_DIR}/config/,file://${REPO_PATH}/examples/sft/config/]"
+)
+CMD_BASE="python ${SRC_FILE} --config-path ${SCRIPT_DIR}/config/ --config-name ${CONFIG_NAME}"
+echo "${CMD_BASE} ${HYDRA_ARGS[*]} ${EXTRA_ARGS}" > "${LOG_FILE}"
+${CMD_BASE} "${HYDRA_ARGS[@]}" ${EXTRA_ARGS} 2>&1 | grep -v "libdav1d" | tee -a "${LOG_FILE}"
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply examples/offline_rl/advantage_labeling/steam/run_steam_sft.sh:43]: Resolved — the Python exit code is propagated via `${PIPESTATUS[0]}`.

File: rlinf/models/embodiment/value_model/steam/modeling_critic.py [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+            probs=probs,
+            atoms=None,
+            hidden_states=hidden_states,
+            progress_values=progress_values,
+        )
+
+    @torch.no_grad()
+    def predict_value(self, observation) -> Tensor:
+        """Return P(progress) per sample (shape ``[B]``)."""
+        return self.predict(observation).predicted_values
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/models/embodiment/value_model/steam/modeling_critic.py:488]: Resolved — docstring fixed to say the signed value in `[-1, 1]` (not `P(progress)`).

File: rlinf/utils/logging.py [hunk 7]
```diff
@@ -18,3 +18,81 @@ def get_logger():
     from rlinf.scheduler.worker import Worker
 
     return Worker.logger
+
+
+def silence_libav_logs() -> None:
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/utils/logging.py:26]: Resolved — made idempotent via a module-level flag.

File: rlinf/models/embodiment/value_model/value_common/image_text_processing.py [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+        do_augment: bool = True,
+        **kwargs,
+    ):
+        if image_processor is None:
+            cls = self._default_image_processor_cls
+            image_processor = (
+                cls(image_keys=image_keys, do_augment=do_augment)
+                if image_keys
+                else cls(do_augment=do_augment)
+            )
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/models/embodiment/value_model/value_common/image_text_processing.py:508]: Resolved — now raises a clear `ValueError` when `_default_image_processor_cls` is unset.

**Comment by Lin-xs** [comment]
1. This PR requires documentation. Please move all of the leading usage examples from the top of each file into the official documentation.
2. Please avoid adding too many model-specific code snippets under `examples/`. If they are related to the supported STEAM model, please place them within the corresponding files under `rlinf/` instead.

File: rlinf/data/process/mixture_config.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+from typing import Any, Union
+
+import yaml
+
+PathLike = Union[str, Path]
+
+
+def mixture_config_path(dataset_path: PathLike) -> Path:
+    """Return ``<dataset_path>/meta/mixture_config.yaml``."""
+    return Path(dataset_path) / "meta" / "mixture_config.yaml"
```
Comment (andylin-hao) [inline rlinf/data/process/mixture_config.py:35]: Two places are not updated to use this new `/meta` subdir:
1. `recap/process/relabel_advantages.py:94` (`dp / "mixture_config.yaml"`), :123, and its own `update_mixture_config write` at `:532` (root);
2. `recap/process/visualize_advantage_dataset.py:851-852` (`config_dir / "mixture_config.yaml"` over `[dataset_path, dataset_path.parent]`). Result: after Step 3, threshold-relabeling and advantage visualization silently miss the threshold/global_return_min/max that compute_advantages wrote and fall back to stats.json/inferred values — different numbers, no error.

File: rlinf/workers/sft/fsdp_steam_sft_worker.py [hunk 10]
```diff
@@ -0,0 +1,1073 @@
...
+import os
+from pathlib import Path
+from typing import Any
+
+os.environ["LIBAV_LOG_LEVEL"] = "quiet"
+os.environ["OPENCV_LOG_LEVEL"] = "OFF"
+logging.getLogger("libav").setLevel(logging.ERROR)
+logging.getLogger("av").setLevel(logging.ERROR)
+
+logger = logging.getLogger(__name__)
```
Comment (andylin-hao) [inline rlinf/workers/sft/fsdp_steam_sft_worker.py]: Use get_logger of rlinf.utils

File: rlinf/hybrid_engines/fsdp/fsdp_model_manager.py [hunk 11]
```diff
@@ -539,11 +539,20 @@ def build_optimizer(
...
-            eps=adam_eps,
-            weight_decay=weight_decay,
-        )
+
+        try:
+            optimizer = torch.optim.AdamW(
+                param_groups,
+                eps=adam_eps,
+                weight_decay=weight_decay,
+                fused=torch.cuda.is_available(),
```
Comment (andylin-hao) [inline rlinf/hybrid_engines/fsdp/fsdp_model_manager.py]: Worker.torch_platform.is_available()

File: rlinf/workers/sft/fsdp_steam_sft_worker.py [hunk 12]
```diff
@@ -0,0 +1,1073 @@
...
+    forward ``model(observation=obs, labels=labels)``.
+    """
+
+    def __init__(self, cfg: DictConfig):
+        Worker.__init__(self)
+        super().__init__(cfg.actor, self._world_size, self._rank)
+
+        self.cfg = cfg
+        torch.cuda.set_device(int(os.environ.get("LOCAL_RANK", 0)))
+        self.device = torch.cuda.current_device()
```
Comment (andylin-hao) [inline rlinf/workers/sft/fsdp_steam_sft_worker.py]: self.torch_platform.set_device
self.torch_platform.current_device()

File: rlinf/workers/sft/fsdp_steam_sft_worker.py [hunk 13]
```diff
@@ -0,0 +1,1073 @@
...
+                    batch_metrics: list[dict[str, float]] = []
+                    for batch in loader:
+                        observation, labels = self._prepare_input(batch)
+                        metrics = self._eval_batch(observation, labels, ensemble_size)
+                        batch_metrics.append(metrics)
+                    if not batch_metrics:
+                        continue
+                    per_dataset[ds_name] = self._mean_metrics(batch_metrics)
+
+            if not per_dataset:
```
Comment (andylin-hao) [inline rlinf/workers/sft/fsdp_steam_sft_worker.py]: The allreduce below will hang if a rank returns here

**Review by andylin-hao — APPROVED** [review]

**Review by guozhen1997 — APPROVED** [review]
